neuralmagic/tests/test_basic_correctness.py

```diff
@@ -0,0 +1,124 @@
+        # ("mistralai/Mixtral-8x7B-Instruct-v0.1", 4096, None),
+        # ("neuralmagic/zephyr-7b-beta-marlin", 4096, None),
+        # ("neuralmagic/OpenHermes-2.5-Mistral-7B-pruned50",
+        #  4096, "sparse_w16a16"),
+        # "NousResearch/Llama-2-7b-chat-hf",
+        # "neuralmagic/TinyLlama-1.1B-Chat-v1.0-marlin",
+        # ("neuralmagic/Llama-2-7b-pruned70-retrained-ultrachat",
+        #  4096, "--sparsity sparse_w16a16"),
+        # "HuggingFaceH4/zephyr-7b-gemma-v0.1",
+        # ("Qwen/Qwen1.5-7B-Chat", 4096, None),
+        # ("microsoft/phi-2", 2048, None),
+        # ("neuralmagic/phi-2-super-marlin", 2048, None),
+        # ("neuralmagic/phi-2-pruned50", 2048, "sparse_w16a16"),
+        # ("mistralai/Mixtral-8x7B-Instruct-v0.1", 4096, None),
+        # ("Qwen/Qwen1.5-MoE-A2.7B-Chat", 4096, None),
+        # ("casperhansen/gemma-7b-it-awq", 4096, None),
+        # ("TheBloke/Llama-2-7B-Chat-GPTQ", 4096, None),
+    ])
+@pytest.mark.parametrize("max_tokens", [32])
+@pytest.mark.parametrize("num_logprobs", [3])
+@pytest.mark.asyncio
+async def test_models_on_server(
+    hf_runner_nm,
+    client,
+    model: str,
+    max_model_len: int,
+    sparsity: str,
+    max_tokens: int,
+    num_logprobs: int,
+) -> None:
+
+    ds = load_dataset("nm-testing/qa-chat-prompts", split="train_sft")
+    example_prompts = [m[0]["content"] for m in ds["messages"]]
+    hf_model = hf_runner_nm(model)
+    hf_outputs = hf_model.generate_greedy_logprobs_nm_use_tokens(
+        example_prompts, max_tokens, num_logprobs)
+
+    del hf_model
+
+    logger = logging.Logger("vllm_server")
+    api_server_args = {
+        "--model": model,
+        "--max-model-len": max_model_len,
+        "--disable-log-requests": None,
+    }
+    if sparsity:
+        api_server_args["--sparsity"] = sparsity
+
+    with VllmServer(api_server_args, logger):
+        completion = await client.completions.create(model=model,
+                                                     prompt=example_prompts,
+                                                     max_tokens=max_tokens,
+                                                     temperature=0.0,
+                                                     logprobs=num_logprobs)
+
+    vllm_outputs = []
+    for req_output in completion.choices:
+        output_str = req_output.text
+        output_tokens = req_output.logprobs.tokens
+        output_logprobs = req_output.logprobs.top_logprobs
+        vllm_outputs.append((output_tokens, output_str, output_logprobs))
+
+    # loop through the prompts
+    check_logprobs_close(
+        outputs_0_lst=hf_outputs,
+        outputs_1_lst=vllm_outputs,
+        name_0="hf_model",
+        name_1="vllm_model",
+    )
+
+    # now repeat using two gpus
+    # specifically doing it here, rather than as a pytest param,
+    # to avoid repeating the huggingface inference and data collection
+    api_server_args["--tensor-parallel-size"] = 2
```

Comment:
With our current server helper implementation, this needs to be a string (the args dict currently must be all strings for names and values). The helper class should actually raise here due to this value not being a string, if it’s not then something is going wrong in the helper…

Reply:
the test execution hasn't gotten this far yet because of the issue w/ spaces in the tokens.  I'll update the value.

Reply:
That makes sense, but even in the initial definition of ``, there are non-string values which should be causing the `VllmServer` class to raise a `ValueError` while processing the dict.

Reply:
Should this --tensor-parallel-size be treated similarly as pytest.mark.parametrize for other arguments e.g. model etc? There is no need to repeat the test code here just for --tensor-parallel-size=2. Also it can be specified to different value than 2 based on the gpu type.

Reply:
I made an execution error, so I didn't actually test this.  I'll pass in a string.

Reply:
@dhuangnm , the problem w/ passing `tensor-parallel-size` as a param is that it would result in the execution of the  code that collects the huggingface data twice.  this is expensive in execution time.

Maybe we could create a session scoped fixture that stores the results the first time, and retrieves the results afterward?

As far as wanting to set the `tensor-parallel-size` arg to a different type based on gpu type, if there's a way to determine that programmatically, we could create a fixture for that.

Reply:
> Maybe we could create a session scoped fixture that stores the results the first time, and retrieves the results afterward?

This also would not survive `pytest-forked`

Reply:
There is a potential solution in the [pypi doc for the pytest-xdist-forked](https://pypi.org/project/pytest-xdist-forked/) package.  see the section **Making session-scoped fixtures execute only once**.  They basically suggest "using a lock file for inter-process communication".  (I'm assuming that it should work w/ the plain pytest-xdist package, but I don't know that.)

Reply:
I think we need to think of a better way for the parameters that can be passed to the server which we may want to test but are not needed to get the HF results. It's not ideal to repeat test code for each parameter.

Reply:
`torch.cuda.device_count` may work for our purposes

Reply:
This conversation is basically moot now that we've decided to have the server and clients run in the same test.